docs/tuplets.md

```diff
@@ -0,0 +1,123 @@
+# Tuplet Design
+
+## Groupings
+Most of then you see a grouping of 3, 5, 6, 7, or 9. However, You may also see a grouping of 2, 4, or 8 if you are a time signature where the top number is an odd number. Also, most places that go over tuplets list the following groupings:
+- Duplet (2)
+- Triplet (3)
+- Quadruplet (4)
+- Quintuplet (5)
+- Sextuplet (6)
+- Septuplet (7)
+- Octuplet (8)
+- Nontuplet (9)
+
+But they all end with etc. Therefore, I don't think we should validate the number. It seems that it can really be any arbitrary number.
+
+## Definition
+A complete tuplet definition requires knowing how many of what duration fits into the space of how many of a certrain duration. This is how you define the ratio.
+You can see a simple dialog box that shows this very clearly [here](https://usermanuals.finalemusic.com/Finale2014Mac/Content/Finale/TPDLG.htm).
+
+Usually the two note durations would be the same, but you can do the calculation with any duration. For instance, you would normally define a standard eighth note triplet in 4/4 as "3 eighth notes in the space of 2 eighth notes". However, you could also define that as "3 eighth notes in the space of 1 quarter note". The ratio here would be 3:2.
+
+## Standard Ratios
+There are certain ratios that are standard, like a triplet would usually have a ratio of 3:2. However, there are also non-standard ratios where you can use whatever you'd like, such as 8:5. It would be good to have the standard ratios defined so that, for instance, a user can create a triplet and not have to define the full ratio.
+
+There should only be a need to list the standard ratios for 2-9. In this case, we can just list them in `Tuplet` struct. This way, if the second number is not specific in the initializer and the first number is one of these, we can fill in the second number according to th static list of default ratios.
+
+## Compound Tuplets
+You can have a tuplet that is made up of another tuplet, which is made up of another tuplet, and so on. 
+
+In order to support this behavior, the `Tuplet` will store an array of type `NoteCollection` so that it can have a `Note` or `Tuplet` as a value in the array. Also, the `NoteCollection` will have to have a property or set of properties to communicate the duration and number of that duration for each `NoteCollection`. 
+    
+i.e. A 3:2 Eighth note tuplet will have a number of 2 and duration of `.eighth`. A single eighth note, would have a number of 1 and duration of `.eighth`
+
+## Validation
+A tuplet needs to always be full. This means if it is a 3:2 ratio, it needs to have 3 notes in it. However, the notes can be rests. Because of this, the only mutating function should be to replace a note, not remove or add.
```

Comment:
I think we need to talk about what to do if you have a Tuplet and want to convert it into single notes maybe as well. 

We do have to have the ratio info in the convert notes in range to Tuplet method. I see that as the main use case for making a Tuplet. 

The duration for the Tuplet (if it's an eighth note triplet vs quarter note triplet) could be mutable. That wouldn't cause a problem. So I'm ok with having that be mutable. As you said, just makes it a little easier to work with probably.